### DISCUSSION TITLE
Compiling without Opengl/glew?

### DISCUSSION THREAD

**Discussion by Obsnold** [smcameron/space-nerds-in-space#304] [General]
Hello!
I've been working on a little learning project trying to get Space Nerds in Space compiling with Yocto for a few tablet like devices I have laying around.
I've managed to get pretty far but the chipset I am using (rk3126) does not have support for glx natively. I tried using the mesa drivers but it caused a lot of conflicts.
I saw the WITHOUTOPENGL flags but it didn't seem to make much difference.

My plan is to use the tablets just for NAVIGATION, ENGINEERING, DAMAGE CONTROL, SCIENCE, and COMMS.

Basically how hard do you think it would be for certain clients to run without opengl? is it possible to just use SDL2?

**Comment by smcameron** [comment]
OpenGL is currently required, I'm afraid.  A long time ago you could compile without OpenGL and it would use my own software renderer (which wasn't very good).  This was back when the game still used GTK.  But I converted it to use SDL2, and dropped GTK, and now OpenGL is required.  Even back then, some screens did not work well without opengl (e.g. MAIN VIEW, WEAPONS, DEMON).

Consider that a Raspberry Pi 4 is too slow for those screens as well.

If you want to use SDL2 without OpenGL (possibly reasonable for comms, engineering, damage control, navigation, science), then you would need to write graph_dev_sdl2.c (start by copying graph_dev_gdk.c, which is the old gtk based code), and link it in instead of graph_dev_opengl.c.  graph_dev.h is the header that defines the interface, and graph_dev_gdk.c and graph_dev_opengl.c are the gdk and opengl implementations, respectively.

You might also have to change snis_graph.c as well to add something to struct snis_graph_context that SDL2 might need.  I'm pretty sure graph_dev_gdk.c doesn't compile anymore because some stuff it needed in struct snis_graph_context was removed when the conversion to SDL2 was done (among possibly many other reasons.)

Also would need to change snis_client.c, in main(), where it sets up OpenGL, maybe move the opengl code into graph_dev_opengl.c and the equivalent non-opengl sdl2 setup into graph_dev_sdl2.c

Hope that helps.

**Comment by Obsnold** [comment]
Thanks for the Help!
In the end I decided it was easier to try and get the mali drivers working.
I had to cut out some of the texture and mesh loading as well due to my device not having much memory.
Got it to the main screen though!

![Untitled](https://user-images.githubusercontent.com/44430222/132809732-e4efba9f-a309-4836-9e00-38ab757a5216.png)

going to set up a server in a bit to tests if it gets any further!

Thanks Again!

**Comment by Obsnold** [comment]
Removing the resources caused other issues so I ended up making a big swap partition.
It takes a while to load and it runs nav just barely. 

![IMG_3611](https://user-images.githubusercontent.com/44430222/132833079-33965d9a-6906-437c-b075-bf420103223e.jpg)
![IMG_3609](https://user-images.githubusercontent.com/44430222/132833086-c497917a-a190-4276-a632-6231916b854e.jpg)

I'll post up my yocto layer in a bit just in case anyone is interested.

**Comment by smcameron** [comment]
Neat!  Although... your FPS of 0.09 seems ... not great.   

FWIW, I have (unrealized) plans to build physical consoles for some stations using Raspberry Pi 4s + some cheap amps + speakers from aliexpress + arduinos + bunch of pots and switches and LEDs and 3D printed aviation style switch things.  Some details here:

* https://github.com/smcameron/snis-consoles
* https://spacenerdsinspace.com/snis-consoles/notes.html

I need to get back on that project... just not very good at making physical stuff.

**Comment by Obsnold** [comment]
Awesome looks good!

Here's a link to the Yocto layer if anyone ever want to see it:
[Yocto_SNIS](https://github.com/Obsnold/Yocto_SNIS/blob/master/README.md)

I wont have as much time to look at it this week but If I can I'll get a setup with a bunch of screens. 
Hopefully I can at least take a nice picture even if it's not playable haha.

**Reply by Obsnold** [reply]
I managed to get a few more devices set up. I think COMMS was running best at about 2.4fps.
Got far enough to get myself killed! Which I consider a success!

![image](https://user-images.githubusercontent.com/44430222/134467426-ab7da289-e01f-4c11-b427-ebb4f14f15e4.jpeg)
![image](https://user-images.githubusercontent.com/44430222/134467460-8c8a5eb9-a7cc-4e2e-96fc-66d432343d6e.jpeg)
![image](https://user-images.githubusercontent.com/44430222/134467520-48aa10a0-9c34-4f46-bda1-f45ef42ac275.jpeg)

**Comment by smcameron** [comment]
Looks cool!

**Comment by smcameron** [comment]
I don't know if you're still messing around with this or not, but in case you are, I just did a little experiment which might help.  The planet textures are most likely what's causing you trouble, as they are quite large.  As an experiment, I tried scaling them down to 128x128 which is *much* smaller than is usual (1024x1024).  They looked surprisingly good at that resolution, so it's something you might try.

See: https://github.com/smcameron/space-nerds-in-space/discussions/333

**Comment by smcameron** [comment]
There is now a --no-textures option of snis-client which may be useful for some stations that don't use textures (e.g.: NAV, ENGINEERING, DAMAGE CONTROL, COMMS, SCIENCE, etc.) which should save ~400Mb of texture memory.  See #334  Also available in the options menu of the snis_launcher script.

**Comment by Obsnold** [comment]
Thanks those changes should help a lot!
I'm no longer working at that company any more so don't have access to all those devices ( They might send me a box of rubbish if I ask nicely).

I have a hankering to build a cockpit again so I might look at using some raspberry pis like you mentioned before. 
I've got an old joystick so now I just want a big panel of switches I can flick.
